packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -138,87 +89,106 @@ library LibProving {
-            fc = blk.forkChoices[fcId];
-            if (
-                fc.blockHash == evidence.blockHash &&
-                fc.signalRoot == evidence.signalRoot &&
-                fc.gasUsed == evidence.gasUsed
-            ) revert L1_SAME_PROOF();
         } else {
-            // This is the branch provers trying to overwrite
-            fc = blk.forkChoices[fcId];
-            if (fc.prover != address(0) && fc.prover != address(1))
-                revert L1_ALREADY_PROVEN();
-
-            if (
-                fc.blockHash != evidence.blockHash ||
-                fc.signalRoot != evidence.signalRoot ||
-                fc.gasUsed != evidence.gasUsed
-            ) revert L1_INVALID_PROOF_OVERWRITE();
+            revert L1_ALREADY_PROVEN();
         }
 
         fc.blockHash = evidence.blockHash;
         fc.signalRoot = evidence.signalRoot;
         fc.gasUsed = evidence.gasUsed;
         fc.prover = evidence.prover;
+        fc.provenAt = uint64(block.timestamp);
 
-        if (evidence.prover == address(1)) {
-            fc.provenAt = uint64(
-                block.timestamp.max(blk.proposedAt + config.proofTimeTarget)
-            );
-        } else {
-            fc.provenAt = uint64(block.timestamp);
+        // Put together the input for proof and signature verification
+        uint256[9] memory inputs;
+
+        inputs[0] = uint256(
+            uint160(address(resolver.resolve("signal_service", false)))
+        );
+        inputs[1] = uint256(
+            uint160(
+                address(
+                    resolver.resolve(config.chainId, "signal_service", false)
+                )
+            )
+        );
+        inputs[2] = uint256(
+            uint160(address(resolver.resolve(config.chainId, "taiko", false)))
+        );
+
+        inputs[3] = uint256(blk.metaHash);
+        inputs[4] = uint256(evidence.parentHash);
+        inputs[5] = uint256(evidence.blockHash);
+        inputs[6] = uint256(evidence.signalRoot);
+        inputs[7] = uint256(evidence.graffiti);
+        inputs[8] =
+            (uint256(uint160(evidence.prover)) << 96) |
+            (uint256(evidence.parentGasUsed) << 64) |
+            (uint256(evidence.gasUsed) << 32);
+
+        bytes32 instance;
+        assembly {
+            instance := keccak256(inputs, mul(32, 9))
         }
 
-        if (evidence.prover != address(0) && evidence.prover != address(1)) {
-            uint256[9] memory inputs;
-
-            inputs[0] = uint256(
-                uint160(address(resolver.resolve("signal_service", false)))
-            );
-            inputs[1] = uint256(
-                uint160(
-                    address(
-                        resolver.resolve(
-                            config.chainId,
-                            "signal_service",
-                            false
-                        )
+        // config.proofTypeEnabled can support 8 different proofs on 8 bits
```

Comment:
The plan is still to abstract the actual implementations away in the verifier, or do you think a simple for loop with some ifs would also work? Seems okay to me like this, but I'm mostly concerned about the additional data needed for the different proof types, having a separate contract for each allows things to be very flexible with what a new proof type needs

Reply:
Yep, agree, thanks for the remark! Outsourced to new files now.